compiler/rustc_codegen_ssa/src/sir.rs

```diff
@@ -341,102 +289,292 @@ impl SirFuncCx<'tcx> {
+        self.push_stmt(bb, stmt);
+        dest_ip
+    }
+
+    fn lower_rvalue<Bx: BuilderMethods<'a, 'tcx>>(
+        &mut self,
+        bx: &Bx,
+        bb: ykpack::BasicBlockIndex,
+        dest_ty: Ty<'tcx>,
+        rvalue: &mir::Rvalue<'tcx>,
+    ) -> ykpack::IPlace {
         match rvalue {
-            mir::Rvalue::Use(opnd) => ykpack::Rvalue::Use(self.lower_operand(opnd)),
-            mir::Rvalue::BinaryOp(op, opnd1, opnd2) => self.lower_binop(*op, opnd1, opnd2, false),
+            mir::Rvalue::Use(opnd) => self.lower_operand(bx, bb, opnd),
+            mir::Rvalue::Ref(_, _, p) => self.lower_ref(bx, bb, dest_ty, p),
+            mir::Rvalue::BinaryOp(op, opnd1, opnd2) => {
+                self.lower_binop(bx, bb, dest_ty, *op, opnd1, opnd2, false)
+            }
             mir::Rvalue::CheckedBinaryOp(op, opnd1, opnd2) => {
-                self.lower_binop(*op, opnd1, opnd2, true)
+                self.lower_binop(bx, bb, dest_ty, *op, opnd1, opnd2, true)
             }
-            mir::Rvalue::Ref(_, _, place) => self.lower_ref(place),
-            mir::Rvalue::Len(place) => ykpack::Rvalue::Len(self.lower_place(place)),
-            mir::Rvalue::Cast(mir::CastKind::Misc, op, ty) => self.lower_cast_misc(op, ty),
-            _ => ykpack::Rvalue::Unimplemented(with_no_trimmed_paths(|| {
+            mir::Rvalue::Cast(mir::CastKind::Misc, op, ty) => self.lower_cast_misc(bx, bb, op, ty),
+            _ => ykpack::IPlace::Unimplemented(with_no_trimmed_paths(|| {
                 format!("unimplemented rvalue: {:?}", rvalue)
             })),
         }
     }
 
-    pub fn lower_place(&self, place: &mir::Place<'_>) -> ykpack::Place {
-        ykpack::Place {
-            local: self.lower_local(place.local),
-            // FIXME projections not yet implemented.
-            projection: place.projection.iter().map(|p| self.lower_projection(&p)).collect(),
+    fn monomorphize<T>(&self, value: &T) -> T
+    where
+        T: TypeFoldable<'tcx> + Copy,
+    {
+        if let Some(substs) = self.instance.substs_for_mir_body() {
+            self.tcx.subst_and_normalize_erasing_regions(substs, ty::ParamEnv::reveal_all(), value)
+        } else {
+            self.tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), *value)
         }
     }
 
-    pub fn lower_projection(&self, pe: &mir::PlaceElem<'_>) -> ykpack::Projection {
-        match pe {
-            mir::ProjectionElem::Field(field, ..) => ykpack::Projection::Field(field.as_u32()),
-            mir::ProjectionElem::Deref => ykpack::Projection::Deref,
-            mir::ProjectionElem::Index(local) => {
-                ykpack::Projection::Index(self.lower_local(*local))
-            }
-            _ => ykpack::Projection::Unimplemented(format!("{:?}", pe)),
-        }
+    /// Wrapper for bx.layout_of() which ensures the type is first monomorphised.
+    fn layout_of<Bx: BuilderMethods<'a, 'tcx>>(&self, bx: &Bx, t: Ty<'tcx>) -> TyAndLayout<'tcx> {
```

Comment:
This function feels like the sort of hack that might confuse us: it has the same name as another function that doesn't do the same thing. Is it so bad to remove this and just say `layout_of(self.monomorphise(...))` when necessary? Do we even need to monomorphise everything?

Reply:
We do need to monomorphise everything, and tracking down missing monomorphisations is (and was) a pain. Hence this function.

Reply:
Let's give it a different name then, something like `monomorphise_and_layout` maybe?

Reply:
Fixed in 1bd0a0